### ISSUE TITLE
Activity com.auth0.android.provider.AuthenticationActivity has leaked ServiceConnection com.auth0.android.provider.CustomTabsController

### ISSUE THREAD

**Issue by Belka1000867** [auth0/Auth0.Android#461]
### Describe the problem

```
ActivityThread: Activity com.auth0.android.provider.AuthenticationActivity has leaked ServiceConnection com.auth0.android.provider.CustomTabsController@4a42179 that was originally bound here
    android.app.ServiceConnectionLeaked: Activity com.auth0.android.provider.AuthenticationActivity has leaked ServiceConnection com.auth0.android.provider.CustomTabsController@4a42179 that was originally bound here
        at android.app.LoadedApk$ServiceDispatcher.<init>(LoadedApk.java:1805)
        at android.app.LoadedApk.getServiceDispatcherCommon(LoadedApk.java:1677)
        at android.app.LoadedApk.getServiceDispatcher(LoadedApk.java:1656)
        at android.app.ContextImpl.bindServiceCommon(ContextImpl.java:1721)
        at android.app.ContextImpl.bindService(ContextImpl.java:1650)
        at android.content.ContextWrapper.bindService(ContextWrapper.java:705)
        at androidx.browser.customtabs.CustomTabsClient.bindCustomTabsService(CustomTabsClient.java:80)
        at com.auth0.android.provider.CustomTabsController.bindService(CustomTabsController.java:72)
        at com.auth0.android.provider.AuthenticationActivity.launchAuthenticationIntent(AuthenticationActivity.kt:70)
        at com.auth0.android.provider.AuthenticationActivity.onResume(AuthenticationActivity.kt:46)
        at android.app.Instrumentation.callActivityOnResume(Instrumentation.java:1446)
        at android.app.Activity.performResume(Activity.java:7939)
        at android.app.ActivityThread.performResumeActivity(ActivityThread.java:4195)
        at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:4237)
        at android.app.servertransaction.ResumeActivityItem.execute(ResumeActivityItem.java:52)
        at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:176)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2016)
        at android.os.Handler.dispatchMessage(Handler.java:107)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7356)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)

```
### Reproduction
Happened after updated to version 2.0.0 

- **Version of this library used:** 2.0.0

**Comment by lbalmaceda** [comment]
@Belka1000867 thanks for reporting this. Can you add some steps to reproduce it? We already have in place [this logic](https://github.com/auth0/Auth0.Android/blob/main/auth0/src/main/java/com/auth0/android/provider/AuthenticationActivity.kt#L57-L63) to remove the reference after the activity is no longer visible and wonder if we should move this logic somewhere else

**Comment by Belka1000867** [comment]
It happens after normal authentication process on `webAuthBuilder.start()`

After debugging I see that `CustomTabsController.unbindService()` does not unbind the service, because `isBound` field is always false. On `CustomTabsController.bindService()` while `CustomTabsClient.bindCustomTabsService(context, preferredPackage, this);` returns false.

**Comment by lbalmaceda** [comment]
thanks for the extra details 🙇  I'll try to reproduce locally and come back to you.

**Comment by lbalmaceda** [comment]
I've tried to reproduce this on an emulator with the Chrome browser app installed. I didn't run into this, in fact, at the time the service should be unbound, the variable `isBound` has a truthy value, so the right logic is executed. 
![image](https://user-images.githubusercontent.com/3900123/113153733-ad1abf80-9237-11eb-9786-ae024edc84b9.png)

The SDK also [logs at "bind time"](https://github.com/auth0/Auth0.Android/blob/main/auth0/src/main/java/com/auth0/android/provider/CustomTabsController.java#L74) whether it could or not bind a Custom Tabs service. This would fail when there are no Custom Tabs compatible browsers installed. Can you check the output of that line in your logcat? Also, can you provide the browser apps you are using to test this or more information around that environment? Thanks

```bash
com.auth0.sample V/CustomTabsController: Trying to unbind the service
com.auth0.sample V/CustomTabsController: Trying to bind the service
com.auth0.sample V/CustomTabsController: Bind request result: true
com.auth0.sample D/CustomTabsController: CustomTabs Service connected
com.auth0.sample D/CustomTabsController: Launching URI. Custom Tabs available: true
```

**Comment by jemaystermind** [comment]
@lbalmaceda I have encountered the same error. I was using DuckDuckGo browser while Chrome is disabled. After enabling Chrome, the error goes away. Hope this helps.

```
V/CallbackHelper: The Callback URI is: demo://jemaystermind.us.auth0.com/android/com.jemaystermind.auth0playground/callback
D/LogoutManager: Using the following Logout URI: https://jemaystermind.us.auth0.com/v2/logout?auth0Client=eyJuYW1lIjoiQXV0aDAuQW5kcm9pZCIsImVudiI6eyJhbmRyb2lkIjoiMzAifSwidmVyc2lvbiI6IjIuMS4wIn0%3D&returnTo=demo%3A%2F%2Fjemaystermind.us.auth0.com%2Fandroid%2Fcom.jemaystermind.auth0playground%2Fcallback&client_id=NGBqZyZD9PKWtsUz11DP8eb1kQehoDV5
V/CustomTabsController: Trying to bind the service
V/CustomTabsController: Bind request result: false
D/CustomTabsController: Launching URI. Custom Tabs available: false
D/Main: onSuccess: User has been logged out!
V/CustomTabsController: Trying to unbind the service
E/ActivityThread: Activity com.auth0.android.provider.AuthenticationActivity has leaked ServiceConnection com.auth0.android.provider.CustomTabsController@963bdb4 that was originally bound here
    android.app.ServiceConnectionLeaked: Activity com.auth0.android.provider.AuthenticationActivity has leaked ServiceConnection com.auth0.android.provider.CustomTabsController@963bdb4 that was originally bound here
        at android.app.LoadedApk$ServiceDispatcher.<init>(LoadedApk.java:1835)
        at android.app.LoadedApk.getServiceDispatcherCommon(LoadedApk.java:1707)
        at android.app.LoadedApk.getServiceDispatcher(LoadedApk.java:1686)
        at android.app.ContextImpl.bindServiceCommon(ContextImpl.java:1819)
        at android.app.ContextImpl.bindService(ContextImpl.java:1749)
        at android.content.ContextWrapper.bindService(ContextWrapper.java:756)
        at androidx.browser.customtabs.CustomTabsClient.bindCustomTabsService(CustomTabsClient.java:80)
        at com.auth0.android.provider.CustomTabsController.bindService(CustomTabsController.java:72)
        at com.auth0.android.provider.AuthenticationActivity.launchAuthenticationIntent(AuthenticationActivity.kt:70)
        at com.auth0.android.provider.AuthenticationActivity.onResume(AuthenticationActivity.kt:46)
        at android.app.Instrumentation.callActivityOnResume(Instrumentation.java:1456)
        at android.app.Activity.performResume(Activity.java:8135)
        at android.app.ActivityThread.performResumeActivity(ActivityThread.java:4428)
        at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:4470)
        at android.app.servertransaction.ResumeActivityItem.execute(ResumeActivityItem.java:52)
        at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:176)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2066)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:223)
        at android.app.ActivityThread.main(ActivityThread.java:7660)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
```

**Comment by lbalmaceda** [comment]
Thanks for the additional details. I'll take a look again using a different browser app.

**Comment by sureshimp766** [comment]
any additional details ? am getting the same error

**Comment by lbalmaceda** [comment]
I was able to reproduce this using `DuckDuckGo` browser, thanks @jemaystermind for the tip. 

I can confirm that while this exception is logged as an error in the logcat (automatically by the OS) **it does not affect the functionality of the SDK** and the authentication can be completed without any issues, as seen in the reproduction video below:

https://user-images.githubusercontent.com/3900123/135073115-14b507d2-0986-4b40-9534-825485c322ba.mp4

Following on my previous comments in this thread, I'm sure that we are taking measures to dispose of the service after the user comes back from the browser. We call the `unbindService` method in the activity's `onDestroy`. But I did notice that this only happens when a custom tabs service is _not available_ and thus, the service is never bound. 

Line with this check: https://github.com/auth0/Auth0.Android/blob/main/auth0/src/main/java/com/auth0/android/provider/CustomTabsController.java#L83

We are keeping track of the bound state in that flag, and only when it was bound, we are unbinding the service. This is the problem: We did call `CustomTabsClient#bindCustomTabsService()` so even if that failed, we still need to call unbind. I'm fixing this by removing the boolean check. See a video with the fix:

https://user-images.githubusercontent.com/3900123/135073946-3874b773-e290-4e02-b7f5-1fdb3d56ea3a.mp4
